[clang][AST] Fix source range of class template implicit instantiations.

frederic-tingaud-sonarsource opened the pull request:
When getting the source range of the implicit instantiation of a class template, we get the source range of the latest declaration of the class template, even when it is not its definition. The problem does not occur with function templates and variable templates.
This patch aligns the behavior of class templates with function and variable templates.

CPP-6863

mizvekov reviewed (commented):
Why is this the correct behavior, rather than the other way around?

The patch seems to leave no way to query the source range of this declaration, rather than the definition.

On the other hand, querying the source range of the definition is currently possible, if you are explicit about it.

frederic-tingaud-sonarsource commented:
Thanks for the feedback @mizvekov. I propose this change for the following reasons:
- Class template implicit instantiations have technically no source range anywhere, which is why I think having the same behavior as function template implicit instantiations or variable template implicit instantiations is the best approach.
- It is also the behavior of other implicit code like special member functions, that are located at the class definition and not at the latest declaration.
- Class template implicit instantiations are definitions: calling `isThisDeclarationADefinition()` on them will return true, so one would expect their range to actually contain a definition.
- Calling `getSourceLocation()` returns the definition already, so there is also a difference of behavior between `getSourceLocation()` and `getSourceRange()` that I find confusing.

mizvekov commented:
> Thanks for the feedback @mizvekov. I propose this change for the following reasons:
> 
> * Class template implicit instantiations have technically no source range anywhere, which is why I think having the same behavior as function template implicit instantiations or variable template implicit instantiations is the best approach.
> * It is also the behavior of other implicit code like special member functions, that are located at the class definition and not at the latest declaration.
> * Class template implicit instantiations are definitions: calling `isThisDeclarationADefinition()` on them will return true, so one would expect their range to actually contain a definition.
> * Calling `ClassTemplateSpecializationDecl::getLocation()` returns the location of the definition already, so there is also a difference of behavior between `getLocation()` and `getSourceRange()` that I find confusing.

Yes, as you point out, template instantiations don't have source locations, as they are not things written by the user. It would make sense to ask what is the source location of the thing it was instantiated from, but that's also confusing when you have multiple declarations.

So I don't think clang has a principled stand here, and I'd like to come up with something that would decide this is the right approach, so if someone in a couple of months comes up with a patch changing everything to the opposite direction, we would be able to decide which is right.

So, what makes this direction seem right to you, why would you even want to ask the source locations of these things, and expect there to be a single answer? Or are you purely concerned about the lack of consistency, and would like a single answer everywhere, regardless of what approach?

If we stopped storing source locations on these things, and stop giving you a single answer, but you would still be able to query what declaration it was instantiated from, and you could decide for yourself which redeclaration to obtain this answer from, would that be okay?

mizvekov commented:
> The most important part for me, is that if `TagDecl::isThisDeclarationADefinition()` returns true, then I need `getSourceRange()` to return the content of the definition.

That's a more interesting limited part of the problem.

It seems strange to me that we produce a template instantiation which is a complete definition, but we somehow think the pattern used to produce it is a non-defining declaration. Maybe we could fix that instead?

frederic-tingaud-sonarsource committed:
CFG: Add a BuildOption to consider default branch of switch on covered enumerations.

By default, the `default:` branch of a switch on fully covered enumarations is considered as "Unreachable".
It is a sane assumption in most cases, but not always. That commit allows to change such behavior when needed.
